src/Symfony/Component/Form/Extension/PasswordHasher/EventListener/PasswordHasherListener.php

```diff
@@ -0,0 +1,80 @@
+<?php
+
+/*
+ * This file is part of the Symfony package.
+ *
+ * (c) Fabien Potencier <fabien@symfony.com>
+ *
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ */
+
+namespace Symfony\Component\Form\Extension\PasswordHasher\EventListener;
+
+use Symfony\Component\Form\Exception\InvalidConfigurationException;
+use Symfony\Component\Form\Extension\Core\Type\RepeatedType;
+use Symfony\Component\Form\FormEvent;
+use Symfony\Component\PasswordHasher\Hasher\UserPasswordHasherInterface;
+use Symfony\Component\PropertyAccess\PropertyAccess;
+use Symfony\Component\PropertyAccess\PropertyAccessorInterface;
+use Symfony\Component\Security\Core\User\PasswordAuthenticatedUserInterface;
+
+/**
+ * @author Sébastien Alfaiate <s.alfaiate@webarea.fr>
+ */
+class PasswordHasherListener
+{
+    private $passwordHasher;
+    private $propertyAccessor;
+
+    private static $passwords = [];
+
+    public function __construct(UserPasswordHasherInterface $passwordHasher, PropertyAccessorInterface $propertyAccessor = null)
+    {
+        $this->passwordHasher = $passwordHasher;
+        $this->propertyAccessor = $propertyAccessor ?? PropertyAccess::createPropertyAccessor();
+    }
+
+    public function registerPassword(FormEvent $event)
+    {
+        $form = $event->getForm();
+        $parentForm = $form->getParent();
+        $mapped = $form->getConfig()->getMapped();
+
+        if ($parentForm && $parentForm->getConfig()->getType()->getInnerType() instanceof RepeatedType) {
+            $mapped = $parentForm->getConfig()->getMapped();
+            $parentForm = $parentForm->getParent();
+        }
+
+        if ($mapped) {
+            throw new InvalidConfigurationException('The hash_property_path option cannot be used on mapped field.');
+        }
+
+        if (!$parentForm || !($user = $parentForm->getData()) || !($user instanceof PasswordAuthenticatedUserInterface)) {
+            throw new InvalidConfigurationException(sprintf('The "hash_property_path" option only supports "%s" objects.', PasswordAuthenticatedUserInterface::class));
```

Comment:
In order to relax this constraint for other cases (like DTOs bound to the form), what do you think about adding support for `PasswordHasherAwareInterface` too?

I think we can support both through the service linked to `PasswordHasherFactoryInterface`.

Reply:
Not sure about this, isn't [`PasswordHasherAwareInterface`](https://symfony.com/doc/current/security/passwords.html#named-password-hashers) supposed to be used only in addition to `PasswordAuthenticatedUserInterface`?

Reply:
Indeed, there is no way to get the string that needs to be hashed with only `PasswordHasherAwareInterface`.

Reply:
Isn't the string to be hashed the `$event->getData()` here? 

If I understand it correctly, `PasswordHasherFactoryInterface` returns a `PasswordHasherInterface` instance which requires a string only to hash, the `$plainPassword`.

Reply:
To be more clear about my suggestion in this thread:
 * You can use `PasswordHasherFactoryInterface` instead of `UserPasswordHasherInterface` to support both `PasswordAuthenticatedUserInterface` and `PasswordHasherAwareInterface`, then you got the proper hasher strategy.
 * Use that hasher service to hash the plain password: `$event->getData()`.

Reply:
As I understand, you suggest supporting password hashing on any type of data?

I would like to consider implementing this in a second time as the current version is covering the most frequent use case and this will require much more work to support this.

Several new options that are currently automatically retrieved:
- an option to choose the hashing method to use
- an option to set the salt if required by the selected hashing method (when using [`LegacyPasswordHasherInterface`](https://github.com/symfony/symfony/blob/c63bd7e455e4833ac52e820497938dc2087f3f42/src/Symfony/Component/PasswordHasher/LegacyPasswordHasherInterface.php#L30)

Reply:
Oh no, my suggestion is simpler than that, no extra options, just take the advantage of the current factory. I will send you a proposal.

> As I understand, you suggest supporting password hashing on any type of data?

Not any type of data, but only instances of `PasswordAuthenticatedUserInterface` OR `PasswordHasherAwareInterface` as you did here.

Reply:
> Not any type of data, but only instances of `PasswordAuthenticatedUserInterface` OR `PasswordHasherAwareInterface` as you did here.

This is what I dont understand.
A class that implements `PasswordHasherAwareInterface` must also implement `PasswordAuthenticatedUserInterface`
So supporting only `PasswordAuthenticatedUserInterface` should be enough.

Reply:
Note also that binding DTOs to forms is recommended over Entities to avoid invalid states. That's why having support for `PasswordHasherAwareInterface` would be a win-win for the community.

Reply:
> A class that implements PasswordHasherAwareInterface must also implement PasswordAuthenticatedUserInterface

Not necessarily, nothing prevents you to use `PasswordHasherAwareInterface` only in a DTOs to change/reset your password without implementing `PasswordAuthenticatedUserInterface`, then you will transfer that hashed password to the final entity which may implement `PasswordAuthenticatedUserInterface` or not.

Reply:
and here `PasswordAuthenticatedUserInterface` does not provide any other value than getting the proper hasher password strategy, and actually, you have 3 options to do that: `string|PasswordHasherAwareInterface|PasswordAuthenticatedUserInterface`.

Reply:
Suggested changes to support `PasswordHasherAwareInterface `:
https://github.com/symfony/symfony/pull/46224/files#r998511474

Reply:
I answered in the related conversation.